staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -219,6 +219,9 @@ type Store struct {
 	StorageVersioner runtime.GroupVersioner
 	// Called to cleanup clients used by the underlying Storage; optional.
 	DestroyFunc func()
+
+	// onceDestroy prevents from Destroy function being called more than once.
+	onceDestroy sync.Once
```

Comment:
This is failing our presubmits, e.g.:
```
# k8s.io/kubernetes/pkg/registry/flowcontrol/flowschema/storage
pkg/registry/flowcontrol/flowschema/storage/storage.go:65:17: assignment copies lock value to statusStore: k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.Store contains sync.Once contains sync.Mutex
```

While this isn't a problem per-se (because we don't call Destroy() for statuses etc.) - it's actually good that the presubmit is failing.
@liggitt - any thoughts on fixing that (we can switch to pointer but then we need to change all places where it is created (i.e. all resources). I bet there have to a better way to do that I'm not seeing...

Reply:
make it a pointer and initialize in CompleteWithOptions()?

Reply:
or, rather than a peer sync.Once, bake a sync.Once into the DestroyFunc when we create it (like we're already doing inside newETCD3Storage)

```diff
diff --git a/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go
index 60ad5a9b6a9..075170f14e4 100644
--- a/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go
+++ b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/storage_factory.go
@@ -70,9 +70,12 @@ func StorageWithCacher() generic.StorageDecorator {
 		if err != nil {
 			return nil, func() {}, err
 		}
+		var once sync.Once
 		destroyFunc := func() {
-			cacher.Stop()
-			d()
+			once.Do(func() {
+				cacher.Stop()
+				d()
+			})
 		}
 
 		// TODO : Remove RegisterStorageCleanup below when PR
diff --git a/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go
index 83c307f7763..4981404faa6 100644
--- a/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go
+++ b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go
@@ -1445,11 +1445,14 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
 		if opts.CountMetricPollPeriod > 0 {
 			stopFunc := e.startObservingCount(opts.CountMetricPollPeriod, opts.StorageObjectCountTracker)
 			previousDestroy := e.DestroyFunc
+			var once sync.Once
 			e.DestroyFunc = func() {
-				stopFunc()
-				if previousDestroy != nil {
-					previousDestroy()
-				}
+				once.Do(func() {
+					stopFunc()
+					if previousDestroy != nil {
+						previousDestroy()
+					}
+				})
 			}
 		}
 	}
```

Reply:
OK - I like the second suggestion - done.